File: app/src/main/java/com/veganbeauty/app/features/home/adapter/HomeBestsellerAdapter.java
package com.veganbeauty.app.features.home.adapter;

import android.graphics.Paint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.veganbeauty.app.data.local.entities.ProductEntity;
import com.veganbeauty.app.databinding.HomeItemBestsellerBinding;

import java.text.NumberFormat;
import java.util.Locale;

public class HomeBestsellerAdapter extends ListAdapter<ProductEntity, HomeBestsellerAdapter.ViewHolder> {

    private final HomeProductCartListener listener;
    private final NumberFormat priceFormatter = NumberFormat.getCurrencyInstance(new Locale("vi", "VN"));

    public HomeBestsellerAdapter(HomeProductCartListener listener) {
        super(new DiffCallback());
        this.listener = listener != null ? listener : new HomeProductCartListener() {
            @Override public void onProductClick(ProductEntity product) {}
            @Override public void onQuickAddToCart(ProductEntity product, View cartButton, android.widget.ImageView productImage) {}
            @Override public void onCartLongPress(ProductEntity product) {}
            @Override public void onBuyNow(ProductEntity product) {}
        };
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        HomeItemBestsellerBinding binding = HomeItemBestsellerBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false
        );
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(getItem(position), position + 1);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        private final HomeItemBestsellerBinding binding;

        public ViewHolder(HomeItemBestsellerBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(ProductEntity product, int rank) {
            View.OnClickListener openDetail = v -> listener.onProductClick(product);
            binding.ivProduct.setOnClickListener(openDetail);
            binding.tvProductName.setOnClickListener(openDetail);
            binding.tvPrice.setOnClickListener(openDetail);
            binding.tvOriginalPrice.setOnClickListener(openDetail);
            binding.tvRank.setOnClickListener(openDetail);

            binding.btnAction.setOnClickListener(v ->
                    listener.onQuickAddToCart(product, binding.btnAction, binding.ivProduct));
            binding.btnAction.setOnLongClickListener(v -> {
                listener.onCartLongPress(product);
                return true;
            });

            binding.tvRank.setText(String.valueOf(rank));
            binding.tvProductName.setText(product.getName());
            binding.tvPrice.setText(priceFormatter.format(product.getPrice()));

            double originalPrice = product.getPrice() / 0.75;
            binding.tvOriginalPrice.setText(priceFormatter.format(originalPrice));
            binding.tvOriginalPrice.setPaintFlags(binding.tvOriginalPrice.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);

            com.veganbeauty.app.utils.ProductImageHelper.loadProductImage(binding.ivProduct, product);
        }
    }

    private static class DiffCallback extends DiffUtil.ItemCallback<ProductEntity> {
        @Override
        public boolean areItemsTheSame(@NonNull ProductEntity oldItem, @NonNull ProductEntity newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull ProductEntity oldItem, @NonNull ProductEntity newItem) {
            return oldItem.equals(newItem);
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/home/adapter/HomeBestsellerAdapter.kt
package com.veganbeauty.app.features.home.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import coil.load
import com.veganbeauty.app.data.local.entities.ProductEntity
import com.veganbeauty.app.databinding.HomeItemBestsellerBinding
import java.text.NumberFormat
import java.util.Locale

class HomeBestsellerAdapter(
    private val onItemClick: (ProductEntity) -> Unit = {}
) : ListAdapter<ProductEntity, HomeBestsellerAdapter.ViewHolder>(DiffCallback()) {

    private val priceFormatter = NumberFormat.getCurrencyInstance(Locale("vi", "VN"))

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = HomeItemBestsellerBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position), position + 1)
    }

    inner class ViewHolder(
        private val binding: HomeItemBestsellerBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(product: ProductEntity, rank: Int) {
            binding.root.setOnClickListener { onItemClick(product) }
            binding.tvRank.text = rank.toString()
            binding.tvProductName.text = product.name
            binding.tvPrice.text = priceFormatter.format(product.price)
            
            val originalPrice = product.price / 0.75 // Assuming 25% discount to match design
            binding.tvOriginalPrice.text = priceFormatter.format(originalPrice)
            binding.tvOriginalPrice.paintFlags = binding.tvOriginalPrice.paintFlags or android.graphics.Paint.STRIKE_THRU_TEXT_FLAG
            
            binding.ivProduct.load(product.mainImage) {
                crossfade(true)
                placeholder(android.R.color.darker_gray)
            }
        }
    }

    private class DiffCallback : DiffUtil.ItemCallback<ProductEntity>() {
        override fun areItemsTheSame(oldItem: ProductEntity, newItem: ProductEntity) =
            oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: ProductEntity, newItem: ProductEntity) =
            oldItem == newItem
    }
}
